File: 21561 - Suvam Tamang/Week 2/1008. Construct Binary Search Tree from Preorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode bstFromPreorder(int[] preorder) {
        TreeNode root = new TreeNode(preorder[0]);   // preorder[0] is anyways the root of the tree by the property of the preorder traversal
        
        for(int i=1; i<preorder.length;i++){
            helperCreate(root, preorder[i]);
        }
        return root;
    }
    
    public void helperCreate(TreeNode root, int value){
        TreeNode temp = root;
        // if the value to be added is lesser than the value that we are currently scanning
        // then we traverse to the left subtree until we find a null value in which case we input the value.
        if(value < temp.val){
            if(temp.left != null){
                helperCreate(temp.left, value);
            }
            else{
                temp.left = new TreeNode(value);
            }
        }
        // if the value to be added is greater than the value that we are currently scanning
        // then we traverse to the right subtree until we find a null value in which case we input the value.
        if(value > temp.val){
            if(temp.right != null && temp.val < value){
                helperCreate(temp.right, value);
            }
            else{
                temp.right = new TreeNode(value);
            }
        }
    }
    
}

File: 21558 - Nithin/week 2/1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
     int idx = 0;
    TreeNode* bstFromPreorder(vector<int>& preorder, int bound = INT_MAX) {
        
        if (idx == preorder.size() || preorder[idx] > bound) 
            return NULL;
        
        TreeNode* root = new TreeNode(preorder[idx++]);
        
        root->left = bstFromPreorder(preorder, root->val);
        
        root->right = bstFromPreorder(preorder, bound);
        
        return root;
    }
};